packages/cubejs-backend-native/src/stream.rs

```diff
@@ -7,76 +7,39 @@ use neon::prelude::*;
 
 use crate::utils::bind_method;
 
-use std::sync::{Condvar, Mutex};
+use tokio::sync::mpsc::{
+    channel as mpsc_channel,
+    error::{TryRecvError, TrySendError},
+    Receiver, Sender,
+};
 
-type JsWriter = Arc<Buffer>;
-type BufferChunk = Result<Option<String>, CubeError>;
+type Chunk = Result<String, CubeError>;
 
 #[derive(Debug)]
-struct Buffer {
-    data: Mutex<Vec<BufferChunk>>,
-    data_cv: Condvar,
-    rejected: Mutex<bool>,
+pub struct StreamReader {
+    receiver: Receiver<Chunk>,
 }
 
-impl Buffer {
-    fn new() -> Self {
-        Self {
-            data: Mutex::new(vec![]),
-            data_cv: Condvar::new(),
-            rejected: Mutex::new(false),
-        }
-    }
-
-    fn push(&self, chunk: BufferChunk) -> bool {
-        if *self.rejected.lock().expect("Can't lock") {
-            return false;
-        }
-
-        let mut lock = self.data.lock().expect("Can't lock");
-        // TODO: check size
-        while lock.len() >= 100 {
-            lock = self.data_cv.wait(lock).expect("Can't wait");
-        }
-        lock.push(chunk);
-        self.data_cv.notify_one();
-
-        true
-    }
-
-    fn release(&self, err: String) {
-        let mut lock = self.rejected.lock().expect("Can't lock");
-        if *lock {
-            return;
-        }
-
-        *lock = true;
-
-        let mut lock = self.data.lock().expect("Can't lock");
-        *lock = vec![Err(CubeError::user(err))];
-        self.data_cv.notify_one();
-    }
-}
-
-impl CubeReadStream for Buffer {
-    fn poll_next(&self) -> BufferChunk {
-        let mut lock = self.data.lock().expect("Can't lock");
-        while lock.is_empty() {
-            lock = self.data_cv.wait(lock).expect("Can't wait");
+impl Iterator for StreamReader {
+    type Item = Chunk;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let poll_wait = std::time::Duration::from_millis(50);
+        loop {
+            match self.receiver.try_recv() {
+                Ok(res) => return Some(res),
+                Err(err) => match err {
+                    TryRecvError::Empty => {
+                        std::thread::sleep(poll_wait);
```

Comment:
`self.receiver.recv().await` will block when it's empty (if i remember correctly). You need to use async API due that you will use it in the Stream. In Tokio you must not use std::thread::sleep. You can use sleep from Tokio, but it's not needed.